OpenAI support (agent sdk, mcp)

planetf1 opened the pull request:
# Tools support for OpenAI Agents

## Summary:
- [X] Add support for internal tools
- [x] Add support for MCP tools

Fixes: #449 
Fixes: #432

## Internal Tools

The Agent SDK supports a limited set of [tools](https://openai.github.io/openai-agents-python/ref/tool/#agents.tool.FunctionTool). Of this list, only the Web Search tool is usable without additional ad-hoc python code (functions), or a vector store (file search), and computer use seems out of scope for maestro currently.

As such I've only implemented web_search. This is very useful for agents, though as with all these tools, it only works when using the new OpenAI Responses API. We only use this if the default URL is left unaltered. ANy custom OpenAI server is assumed (for now) to only support chat completions (for example Ollama)

New test yamls are added (but not automated for now, since they require the OpenAI API key)

### Example Output:

**Workflow prompt:** What are the top news headlines in the United Kingdom today
**Agent instructions:** You are a helpful agent. Respond to the users question, making use of any required tools

```shell
((.venv) ) ➜  maestro git:(issue384_openai4) ✗ maestro run tests/agents/openai_agent/agents_search.yaml tests/agents/openai_agent/workflow_search.yaml
2025-04-24 13:05:45.041 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
🔓 DEBUG [OpenAIAgent openai_test_search]: Using Base URL: https://api.openai.com/v1
🔓 DEBUG [OpenAIAgent openai_test_search]: Using Model: gpt-4o-mini
🔓 DEBUG [OpenAIAgent openai_test_search]: Tool 'web_search' requested.
🔓 INFO [OpenAIAgent openai_test_search]: Added supported tool: web_search
🔓 Running openai_test_search...
🔓 Response from openai_test_search: As of April 24, 2025, here are the top news headlines in the United Kingdom:

**Steel Industry (Special Measures) Act 2025 Enacted**

The UK Parliament has passed the Steel Industry (Special Measures) Act 2025, a significant piece of legislation aimed at preventing the closure of steel production infrastructure. The bill was fast-tracked through Parliament, with both the House of Commons and the House of Lords recalled from recess to debate and approve the legislation. The Act received royal assent on April 12, 2025. ([en.wikipedia.org](https://en.wikipedia.org/wiki/Steel_Industry_%28Special_Measures%29_Act_2025?utm_source=openai))
```

(continues)


## MCP Tools

A list of MCP endpoints is specified in the *MAESTRO_MCP_ENDPOINTS* environment variable
ie
```
export MAESTRO_MCP_ENDPOINTS="http://localhost:8000/sse"                                  
```

These tools must be run independently from maestro & must support http streaming (sse) - not just stdio.

No tools are specified in the yaml - the agent appears to automatically have access to all tools surfaced by the server.

It may be possible to extend the problem to reduce the scope - but as an initial implementation I don't this is a problem:


New examples have been added :

**Workflow prompt:** What tools do you have access to?
**Agent instructions:** You are a helpful agent. Respond to the users question, making use of any required tools

```
((.venv) ) ➜  maestro git:(issue384_openai4) ✗ maestro run tests/agents/openai_agent/agents_mcp.yaml tests/agents/openai_agent/workflow_mcp.yaml
2025-04-24 14:39:43.272 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
🔓 DEBUG [OpenAIAgent openai_test_mcp]: Using Base URL: https://api.openai.com/v1
🔓 DEBUG [OpenAIAgent openai_test_mcp]: Using Model: gpt-4o-mini
🔓 INFO [OpenAIAgent openai_test_mcp]: Using default base URL. Assuming Responses API is available for tools if supported by the model/library.
🔓 DEBUG [OpenAIAgent openai_test_mcp]: No tools specified or 'web_search' not requested.
🔓 DEBUG [OpenAIAgent openai_test_mcp]: Attempting to connect to MCP Servers: ['http://localhost:8000/sse']...
🔓 INFO [OpenAIAgent openai_test_mcp]: MCP Server connected: openai_test_mcp_MCP_Server_1 (http://localhost:8000/sse)
🔓 INFO [OpenAIAgent openai_test_mcp]: Running agent with 1 active MCP server(s).
🔓 Running openai_test_mcp with prompt...
🔓 DEBUG [OpenAIAgent openai_test_mcp]: Agent run completed.
🔓 Response from openai_test_mcp: I have access to a variety of tools to assist with different tasks, including:

1. **Math Calculations**: Performing arithmetic operations, evaluating expressions, and calculating BMI.
2. **Search and Information Retrieval**: Searching for research papers, news articles, Wikipedia content, and more.
3. **GitHub Operations**: Managing GitHub repositories, including creating and listing issues and pull requests.
4. **Weather Information**: Fetching current weather data and forecasts.
5. **File and Image Handling**: Retrieving files from GitHub and creating image thumbnails.

If you have a specific task in mind, feel free to ask!
```

## Test MCP Server

https://github.com/planetf1/mcp-server was run locally.
This server was created for education and isn't intended to really be useful, rather just to understand mcp and help with testing. 

We may need to add a small mcp server to support automated testing

## *NEW* Local MCP Server

Local MCP servers can now be used. For example 
```
export MAESTRO_MCP_ENDPOINTS="http://localhost:8000/sse, /Users/jonesn/bin/github-mcp-server stdio"
```

```
((.venv) ) ➜  maestro git:(issue384_openai4) ✗ maestro run tests/agents/openai_agent/agents_ollama_mcp.yaml tests/agents/openai_agent/workflow_ollama_mcp.yaml
2025-04-25 11:30:17.543 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Base URL: http://localhost:11434/v1
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Model: granite3.3:8b
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: No tools specified or 'web_search' not requested.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Attempting to connect to MCP Servers: ['http://localhost:8000/sse', '/Users/jonesn/bin/github-mcp-server stdio']...
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: MCP Server (SSE) connected: openai_test_mcp_ollama_MCP_Server_1_SSE (http://localhost:8000/sse)
GitHub MCP Server running on stdio
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: MCP Server (Stdio) connected: openai_test_mcp_ollama_MCP_Server_2_Stdio (/Users/jonesn/bin/github-mcp-server stdio)
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Running agent with 2 active MCP server(s).
🔓 Running openai_test_mcp_ollama with prompt...
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Agent run completed.
🔓 Response from openai_test_mcp_ollama: As an assistant, I have access to a variety of information and tools including:

1. Text processing and conversation capabilities.
2. A comprehensive knowledge cutoff of 2021-09 so I can provide up-to-date information until that time.
3. Tools for searching through code across GitHub repositories (search_code).
4. Tools for searching issues, pull requests, users, repositories, and committed code on GitHub (search_issues, search_repositories, search_users).
5. Capabilities for searching the web using an external HTTP request (allowing me access to general knowledge up until my current knowledge base cutoff).
6. Tools for code-related functionalities such as:
   - Push files to a repository in a single commit (push_files).
   - Search for specific code snippets across repositories (search_code).
   - Update issues and pull requests in GitHub repositories (update_issue, update_pull_request_branch).
7. Code execution and manipulation capabilities with tools like execute_code.
8. Code translation between different programming languages (translate_code).
9. Formatting and prettifying code snippets (format_code).
10. Tools for reviewing pull requests on GitHub (review_pull_request).
11. Provide summaries of lengthy texts or code blocks (summarize).
12. Tools to analyze code for complexity, potential bugs, and other issues (analyze_code). 
13. Tools for comparing different versions of files or commits in a repository (compare_files_or_commits).
14. Various project management tools such as extracting issue and comment data from GitHub, listing assigned labels, milestones, projects, and more (list_issues_for_milestone, list_labels, etc.).
15. Generating visualizations or charts from code metrics.
16. Tools to assist in the software development workflow like creating checklists for developers or providing guides on code review practices.

Please note that my actual capabilities heavily depend on the precise functions and data access of the specific AI model implementation I'm based on. The list above is a representation of potential functionalities one might expect from an assistant tool similar to mine, grounded in the provided function descriptions.
```
## Observability

Pydantic's 'logfire' library has been used to automatically instrument the OpenAI library calls, and convert to Open Telemetry. These events can be sent to any Open Telemetry compatible endpoint by setting `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` for example:

```
OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://localhost:4318/v1/traces
```

This gives visability of LLM prompts and responses.

Note: Currently regular debug logging is just printed to stdout/stderr.

## Streaming

Set `MAESTRO_OPENAI_STREAMING=true` to force use of streaming - this is primarily to support test/dev. 
This could be removed once #447 is implemented

## Testing

A new demo has been added under `demos/workflows/openai-mcp.ai`. It's more a technical demo than a user scenario. In addition to a 'dry run' it adds support for a *real* run with live tools & ollama:
* changed test scripts to run a local `test.sh` rather than the common one, if found -- this was needed to setup environment and additional scripts for running a real demo
* test script downloads the `github-mcp-server` ready to be run (locally) by the demo
* test yaml specifies a small model so it runs in sensible time (a few minutes) in a github action
* test.yaml does some trivial parsing of the output to look for 2 known lines
* updated github action to install the latest version of ollama (required for recent models)

## Context Window

`MAESTRO_OPENAI_MAX_TOKENS` can be set to specify the max tokens used by the Agent.
Note: some debug suggests this may not be fully working ... default appears to be 2048

## Known issues & limitations

* In this version a lot of text is sent to stdout - in future this may be better as a debug trace
* The mcp support is for openAI only in this version
* There's no support in the yaml for any configuration etc for mcp
* When multiple mcp servers are specified, it appears that only one server's list of tools is used (being investigated)
* There's no way to limit the environment being passed to the local mcp server. (with the remote server, this is the user's responsibility)
* There's no way to limit the tools being evaluated, except through prompting (too many tools can confuse llms)
* The code evolved, and could benefit from some refactoring
* Structure of test setup, environment, scaffolding, capturing output could do with generalizing further and refactoring.

planetf1 commented:
Surprisingly(!) the MCP support is working ok with ollama+granite3.3:8b too:
```
export OPENAI_API_KEY=ollama
export OPENAI_BASE_URL=http://localhost:11434/v1
```
 
```
((.venv) ) ➜  maestro git:(issue384_openai4) maestro run tests/agents/openai_agent/agents_ollama_mcp.yaml tests/agents/openai_agent/workflow_ollama_mcp.yaml
2025-04-24 15:30:37.562 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Base URL: http://localhost:11434/v1
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Model: granite3.3:8b
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Non-default base URL detected. Forcing 'chat_completions' API. The 'web_search' tool may be unavailable.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: No tools specified or 'web_search' not requested.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Attempting to connect to MCP Servers: ['http://localhost:8000/sse']...
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: MCP Server connected: openai_test_mcp_ollama_MCP_Server_1 (http://localhost:8000/sse)
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Running agent with 1 active MCP server(s).
🔓 Running openai_test_mcp_ollama with prompt...
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Agent run completed.
🔓 Response from openai_test_mcp_ollama: I can perform various tasks using several tools and APIs. Here's a list of functionalities I currently have:

1. `long_task`: This function processes multiple files with progress tracking.
2. Search functions for different search engines:
   - `mojeek_search` for Mojeek search engine.
   - `news_search` for news articles using NewsAPI, allowing querying for specific time frames and sources.
   - `tavily_search` for web searches using Tavily's AI-powered search API with adjustable search depths.
   - `wikipedia_search` to search Wikipedia and retrieve content limits to a specified number of results.
3. Code search on GitHub: `github_search_code`.
4. User activity analysis on GitHub with the `github_user_activity`.
5. Browsing and searching documentation on Mojek, OpenMeteo, and NewsAPI:
   - `mojeek_search`
   - `openmeteo_forecast` for weather forecasts.
   - `news_search` for news articles search within specific time frames and sources.
6. General web searching using Tavily's API.
7. Utilizing functions related to file processing such as:
   - `long_task`: Processes files with progress tracking.
   - Multiple API search functionalities including Mojeek, NewsAPI, Tavily, Wikipedia, and GitHub.
   
Please let me know how I can assist you with these tools.
```

planetf1 commented:
Description updated - works with local MCP servers

I'd propose we consider merging this - even though there's tidy up, tests, docs, debugging to sort, it's a big step in functionality that is really useful (including for cbom) & will not impact non-openAI agents (which we didn't have before)

planetf1 commented:
@aki doc added with more info.

akihikokuroda commented:
Docs look good. Thanks!  Last request.  Would you add some simple demos with ollama under `maestro/demos` so that the demo test pick it up and run every PR?

george-lhj commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,148 @@
+  mode: local # 'local' mode uses this Maestro agent implementation
+```
+
+It is important to set the `OPENAI_API_KEY` environment variable - see [OpenAI's website](https://platform.openai.com/account/api-keys).
````
are we using personal keys currently? or do we have one that we all can use?

planetf1 commented:
> Docs look good. Thanks! Last request. Would you add some simple demos with ollama under `maestro/demos` so that the demo test pick it up and run every PR?

I'd prefer to do that in #449 as there are some complexities with needing a real backend, as well as getting the environment right for tools. This is a first step which shouldn't impact other code. I'll look at the tests next week.

akihikokuroda commented:
@planetf1 I rather wait for the tests.  @george-lhj I'll be out until Thursday next week. Do you need this for improving your demos soon? If so, please try and approve it.  Thanks!

george-lhj commented:
I don't need it for demos right now so I can wait

planetf1 commented:
Refactored the code -- currently both the mcp & observability is openai only. 
 - observability uses the openai api for the 'intercepting' of llm calls. There ae similar approaches for langfuse with other libraries.
  - mcp also uses the mcp server types from openai. Other libraries have similar concepts, though expressed through different types 
  
  Both could probably be generalized, but it would make sense to look at the beeai and crewai support & figure out how to break out the common aspects.

planetf1 commented:
Added `MAESTRO_OPENAI_STREAMING` to allow use of streaming to be forced. Currently the framework doesn't support this, but for completeness I added to the openAI support. this allows testing & could be removed once the framework is in place.

planetf1 commented:
Seeing one warning in debug output
```
time=2025-04-29T18:22:37.749+01:00 level=WARN source=runner.go:131 msg="truncating input prompt" limit=2048 prompt=5131 keep=4 new=2048
```

This is despite setting max_tokens on the Model Properties...

planetf1 commented:
* refactored to clean up
* removed langfuse integration - too complex. solution specific. Simplified to just support sending traces to Open Telemetry endpoint. This provides insight into llm prompts & responses. Additional debug is not yet sent, and this is openai only for now.

planetf1 commented:
@george-lhj Did you get a chance to give this a go?

planetf1 commented:
Current version (running the demo workflow):
```
((.venv) ) ➜  maestro git:(issue384_openai4) maestro run demos/workflows/openai-mcp.ai/agents.yaml demos/workflows/openai-mcp.ai/workflow.yaml
WARNING: Could not import crewai. CrewAI agent support will be disabled. Run `pip install crewai litellm==1.67.0.post1` to enable
2025-04-30 13:53:18.168 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Base URL: http://localhost:11434/v1
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Model: granite3.3:2b
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Using 'chat_completions' API.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: No static tools requested.
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Using max_tokens: 64000
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Set OLLAMA_CONTEXT_LENGTH to 64000
🔓 DEBUG [openai_test_mcp_ollama - MCP Setup]: Attempting MCP connections: ['/Users/jonesn/bin/github-mcp-server stdio']...
GitHub MCP Server running on stdio
🔓 INFO [openai_test_mcp_ollama - MCP Setup]: MCP Server (Stdio) connected: openai_test_mcp_ollama_MCP_Server_1_Stdio (/Users/jonesn/bin/github-mcp-server stdio)
🔓 INFO [openai_test_mcp_ollama - MCP Setup]: Successfully connected to 1 MCP server(s).
🔓 Running openai_test_mcp_ollama with prompt...
12:53:18.244 Chat Completion with 'granite3.3:2b' [LLM]
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Agent run completed.
🔓 Response from openai_test_mcp_ollama: Tools found 12:
1. search_repositories
2. search_users
3. search_code
4. search_issues
5. update_issue
6. update_pull_request_branch
7. search_repositories
8. search_users
9. search_code
10. search_issues
11. update_issue
12. update_pull_request_branch
13. /api/version
14. /check_username
15. /detect_lang
16. /diff
17. /get_user_repos
18. /history
19. /list_issues_and_pull_requests
20. /search_repos
21. /search_users
22. /users/lookup
```

The output varies -- the prompts used with this integration need to guide the LLM more, and use of a larger model should give better responses.

george-lhj commented:
yes i am testing this now

george-lhj reviewed (approved):
I don't have an openai key, so i was not able to test the openai models. However, I swtiched to the llama/granite models which worked for me. 

I think this PR introduces one new library, logfire, which should be added to our poetry file. Also, this would be really great to demo at our team meeting or weekly maestro meeting.

Other than that, lgtm! thanks

george-lhj commented:
Here's the tools from the output I get in llama:

```
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Base URL: http://localhost:11434/v1
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Using Model: llama3.1
🔓 INFO [OpenAIAgent openai_test_mcp_ollama]: Using 'chat_completions' API.
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: No static tools requested.
🔓 DEBUG [openai_test_mcp_ollama - MCP Setup]: No MCP endpoints configured in MAESTRO_MCP_ENDPOINTS.
🔓 Running openai_test_mcp_ollama with prompt...
19:09:18.067 Chat Completion with 'llama3.1' [LLM]
🔓 DEBUG [OpenAIAgent openai_test_mcp_ollama]: Agent run completed.
🔓 Response from openai_test_mcp_ollama: I have access to a wide range of tools that I can utilize to provide accurate and informative responses. Here are some of the categories and specific tools I have at my disposal:

**Information Retrieval**

1. Google Search API: Allows me to perform custom searches on the web.
2. Wolfram Alpha API: Provides access to a vast repository of knowledge, including math, science, and general information.

**Math and Calculation**

1. Python interpreter: Enables me to perform mathematical calculations, data analysis, and algorithm development.
2. Wolfram Alpha calculator: Allows me to perform complex math and science computations.
3. Symbolic mathematics library (e.g., Sympy): Facilitates symbolic manipulation of mathematical expressions.

**Natural Language Processing (NLP)**

1. NLTK: A comprehensive NLP library for text processing, tokenization, and sentiment analysis.
2. spaCy: Another popular NLP library for entity recognition, language modeling, and semantic search.
3. Stanford CoreNLP: A Java library for NLP tasks like named entity recognition, part-of-speech tagging, and dependency parsing.

**Knowledge Graphs**

1. Wikipedia API: Allows access to a vast repository of structured knowledge.
2. ConceptNet API: A large-scale multilingual knowledge graph that encompasses various topics.
3. YAGO (Yet Another Glossary Ontology) database: A comprehensive ontology covering a wide range of domains.

**Time and Date Calculation**

1. Dateutils library (Python): Performs date calculations, formatting, and conversions.
2. Moment.js: A JavaScript library for manipulating dates and times.

**Visualization and Display**

1. Matplotlib (Python): Creates static, animated, or interactive visualizations using plots, charts, and graphs.
2. Seaborn (Python): Provides a high-level interface for drawing statistical graphics.
3. Tableau API: Allows integration with business intelligence tools to create interactive dashboards.

**Other Tools**

1. Random number generator: For generating random numbers, phrases, or characters.
2. Web scraping libraries (e.g., BeautifulSoup, Scrapy): Accessible for extracting data from websites that do not provide APIs.
3. File system access (on certain systems): Enables me to manipulate local files and perform operations like reading and writing.

These tools are used to answer your questions on various topics, including but not limited to:

* Math and science
* Programming languages and development
* Language translation and linguistics
* Business, economics, and finance
* History, geography, and culture
* Health and medicine
* Technology and computing
* Literature and education
```

george-lhj commented:
lmk if you want to make any more changes, or if I should merge it. thanks

planetf1 commented:
> I don't have an openai key, so i was not able to test the openai models. However, I swtiched to the llama/granite models which worked for me.

Thanks for checking

> I think this PR introduces one new library, logfire, which should be added to our poetry file. Also, this would be really great to demo at our team meeting or weekly maestro meeting.

Did you have issues after doing a `poetry sync` or clean install? There is already a dependency 
```
 pydantic-ai = {extras = ["logfire"], version = "^0.1.8"}
```
 which seems to be the recommendation for logfire (I've not tried just 'logfire')

planetf1 commented:
> Here's the tools from the output I get in llama:

> 🔓 DEBUG [openai_test_mcp_ollama - MCP Setup]: No MCP endpoints configured in MAESTRO_MCP_ENDPOINTS.

Did you try MCP? 
I presume not - as it did not find any?

> 🔓 Response from openai_test_mcp_ollama: I have access to a wide range of tools that I can utilize to provide accurate and informative responses. Here are some of the categories and specific tools I have at my disposal:

The 'demo' instructions refer to how to setup an example mcp tool (github mcp)

planetf1 commented:
> lmk if you want to make any more changes, or if I should merge it. thanks

I'm not actively planning any changes right now, so I think we could merge, but given I think both @akihikokuroda and @maximilien are back today, perhaps leave it until later in the day (still morning your time) in case of any objections then please do merge.

Generally, I think it will be appropriate to make further changes/fixes on new PRs

planetf1 commented:
Two more changes
 * MAESTRO_OPENAI_EXTRA_HEADERS : Can use to pass extra headers (ie for some inference servers that require them to authenticate)
 * MAESTRO_OPENAI_USE_LITELLM: The OpenAI agent sdk will only handle models/backends it knows about, for example passing a model `service/llama3.3' may produce an error like `agents.exceptions.UserError: Unknown prefix: service`. The agents SDK now supports litellm as a backend, so this gives a broader reach

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+The full example can be run with:
+
+```shell
+maestro run tests/agents/openai_agent/agents.yaml test/agents/openai_agent/workflow.yaml
````
`test/agents/openai_agent/workflow.yaml` to `tests/agents/openai_agent/workflow.yaml`

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+The full example can be run with:
+
+```shell
+maestro run tests/agents/openai_agent/agents_search.yaml test/agents/openai_agent/workflow_search.yaml
````
same as above

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+An example that assumes Ollama with the `granite3.3:8b model available` can be run with:
+
+```shell
+maestro run tests/agents/openai_agent/agents_search.yaml test/agents/openai_agent/workflow_search.yaml
````
same as above

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+
+For *OpenAI* :
+```shell
+maestro run tests/agents/openai_agent/agents_mcp.yaml test/agents/openai_agent/workflow_mcp.yaml
````
same as above

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+4.  **Run Maestro:**
+
+```shell
+maestro run tests/agents/openai_agent/agents_mcp.yaml test/agents/openai_agent/workflow_mcp.yaml
````
same as above

akihikokuroda commented on maestro/docs/README_openai.md:
````diff
@@ -0,0 +1,201 @@
+```shell
+or with *Ollama*
+```shell
+maestro run tests/agents/openai_agent/agents_ollama_mcp.yaml test/agents/openai_agent/workflow_ollama_mcp.yaml
````
same as above

planetf1 commented:
@akihikokuroda thanks for the review comments.
Fixes are in https://github.com/i-am-bee/beeai-labs/pull/461 ready for review

akihikokuroda commented on maestro/src/agents/openai_mcp.py:
```diff
@@ -0,0 +1,91 @@
+
+# Uses openai agent specific types - though concepts are similar across implementations
+# TODO: can this be refactored so we can support more types of agents
+async def setup_mcp_servers(
```
It's better have some unit tests for this function.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
 
-class OpenAIAgent(Agent):
+
+SUPPORTED_TOOL_NAME: Final[str] = "web_search"
```
Is this the only supported static tool?

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
+                if isinstance(headers_dict, dict):
+                    valid_headers = {str(k): str(v) for k, v in headers_dict.items()}
+                    # don't log: may contain API key (secret)
+                    obfuscated_headers = {k: '*****' for k in valid_headers}
```
It probably simpler to print the keys in the header.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
+                    self.print(f"INFO [OpenAIAgent {self.agent_name}]: Using extra headers: {obfuscated_headers} (values redacted in log)")
+                    return valid_headers
+                else:
+                    self.print(f"WARN [OpenAIAgent {self.agent_name}]: MAESTRO_OPENAI_EXTRA_HEADERS is not a valid JSON dictionary: '{headers_str}'. Ignoring.")
+            except json.JSONDecodeError:
+                self.print(f"WARN [OpenAIAgent {self.agent_name}]: MAESTRO_OPENAI_EXTRA_HEADERS contains invalid JSON: '{headers_str}'. Ignoring.")
+            except Exception as e:
+                 self.print(f"WARN [OpenAIAgent {self.agent_name}]: Error processing MAESTRO_OPENAI_EXTRA_HEADERS='{headers_str}': {e}. Ignoring.")
+        return None
```
The header stings may still have some keys in the error cases.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
+            return final_output_str
+        else:
+            self.print(f"WARN [OpenAIAgent {self.agent_name}]: Agent run completed but no 'final_output' found.")
+            messages = getattr(result, 'messages', [])
+            last_message_content = messages[-1].content if messages and hasattr(messages[-1], 'content') else "No message content available."
```
This may cause an exception.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
+            return fallback_str
+
+
+    # TODO: Cleanup streaming vs non-streaming
```
Please do this.  These 2 functions have enough duplicate code.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
-        result = await OAIRunner.run(self.openai_agent, prompt)        
-        self.print(f"Response from {self.agent_name}: {result.final_output}")
-        return result.final_output
+        streaming_override = os.getenv("MAESTRO_OPENAI_STREAMING", "auto").lower()
```
It seems that this `run` and `run_streaming` can be simplified by using the common code.

akihikokuroda reviewed (commented):
This has been merged already but I put some review comments here for easy reference.
 I added some additional comments

akihikokuroda commented on maestro/src/agents/openai_mcp.py:
```diff
@@ -0,0 +1,91 @@
+    endpoint_definitions = [ep.strip() for ep in mcp_endpoints_str.split(',') if ep.strip()]
+
+    if not endpoint_definitions:
+        print_func(f"DEBUG [{agent_name} - MCP Setup]: No MCP endpoints configured in MAESTRO_MCP_ENDPOINTS.")
```
The debug messages should be disabled by default.

akihikokuroda commented on maestro/src/agents/openai_agent.py:
```diff
@@ -1,87 +1,394 @@
+        self.uses_chat_completions: bool = self.base_url != OPENAI_DEFAULT_URL
+        self.use_litellm: bool = os.getenv("MAESTRO_OPENAI_USE_LITELLM", "false").lower() == "true"
+        self.endpoint_has_tracing: bool = self.base_url == OPENAI_DEFAULT_URL
+        self.print(f"DEBUG [OpenAIAgent {self.agent_name}]: Using Base URL: {self.base_url}")
```
The DEBUG nessages should be disabled by default.

maximilien commented on maestro/demos/workflows/openai-mcp.ai/README.md:
```diff
@@ -0,0 +1,44 @@
+export GITHUB_PERSONAL_ACCESS_TOKEN=<github-token>
+export OPENAI_API_KEY=ollama
+export OPENAI_BASE_URL="http://localhost:11434/v1" 
+export MAESTRO_MCP_ENDPOINTS="/Users/jonesn/bin/github-mcp-server stdio"
```
Do we want to have hardcoded path specific to one user?

maximilien reviewed (commented):
I would love to see if we can simplify this? Thx
